x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields_row.tsx

```diff
@@ -0,0 +1,262 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useMemo } from 'react';
+import {
+  EuiButtonIcon,
+  EuiComboBox,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiFormRow,
+  EuiIcon,
+  EuiTextColor,
+} from '@elastic/eui';
+import { euiThemeVars } from '@kbn/ui-theme';
+import { FIELD_TYPES, UseField } from '../../../../shared_imports';
+import { useNameField } from './use_name_field';
+import { useTypeField } from './use_type_field';
+import * as i18n from './translations';
+
+import type {
+  ArrayItem,
+  ERROR_CODE,
+  FieldConfig,
+  FieldHook,
+  FormData,
+  ValidationFunc,
+} from '../../../../shared_imports';
+import type { RequiredFieldInput } from '../../../../../common/api/detection_engine/model/rule_schema/common_attributes.gen';
+import type { RequiredFieldWithOptionalEcs } from './types';
+
+const SINGLE_SELECTION_AS_PLAIN_TEXT = { asPlainText: true };
+
+interface RequiredFieldRowProps {
+  item: ArrayItem;
+  removeItem: (id: number) => void;
+  typesByFieldName: Record<string, string[] | undefined>;
+  availableFieldNames: string[];
+  getWarnings: ({ name, type }: { name: string; type: string }) => {
+    nameWarning: string;
+    typeWarning: string;
+  };
+  parentFieldPath: string;
+}
+
+export const RequiredFieldRow = ({
+  item,
+  removeItem,
+  typesByFieldName,
+  availableFieldNames,
+  getWarnings,
+  parentFieldPath,
+}: RequiredFieldRowProps) => {
+  const handleRemove = useCallback(() => removeItem(item.id), [removeItem, item.id]);
+
+  const rowFieldConfig: FieldConfig<
+    RequiredFieldWithOptionalEcs,
+    RequiredFieldInput,
+    RequiredFieldInput
+  > = useMemo(
+    () => ({
+      type: FIELD_TYPES.JSON,
+      deserializer: (value) => {
+        const rowValueWithoutEcs: RequiredFieldInput = {
+          name: value.name,
+          type: value.type,
+        };
+
+        return rowValueWithoutEcs;
+      },
+      validations: [{ validator: makeValidateRequiredField(parentFieldPath) }],
+      defaultValue: { name: '', type: '' },
+    }),
+    [parentFieldPath]
+  );
+
+  return (
+    <UseField
+      key={item.id}
+      path={item.path}
+      config={rowFieldConfig}
+      component={RequiredFieldRowContent}
+      readDefaultValueOnForm={!item.isNew}
+      componentProps={{
+        itemId: item.id,
+        onRemove: handleRemove,
+        typesByFieldName,
+        getWarnings,
+      }}
+      availableFieldNames={availableFieldNames}
+    />
+  );
+};
+
+interface RequiredFieldRowInnerProps {
+  field: FieldHook<RequiredFieldInput>;
+  onRemove: () => void;
+  typesByFieldName: Record<string, string[] | undefined>;
+  availableFieldNames: string[];
+  getWarnings: ({ name, type }: { name: string; type: string }) => {
+    nameWarning: string;
+    typeWarning: string;
+  };
+  itemId: string;
+}
+
```

Comment:
<img width="1275" alt="Screenshot 2024-05-08 at 16 18 18" src="https://github.com/elastic/kibana/assets/92328789/b4abb2db-99b6-4a89-bc19-110075055e52">


What's expected behaviour for custom ES|QL created fields? I can see I can add it and then define field type.
But warning still there

Reply:
This behavior is actually expected. 

We allow users to add custom fields, even if the index hasn't been created yet. This could be the case, for example, after installing a prebuilt rule. In such scenario, we  to show a warning to indicate that the field doesn't exist, but still allow to save the rule.

So if you add a custom field, you'll see a warning. But once you actually create a mapping for this field, the warning should be gone.

Reply:
Mapping won't be created for custom field defined in ES|QL query, but will be saved in alert.
Newly created fields are returned only in ES|QL response, so they always be unmapped, unless user create runtime field.

I think we might want to document this

cc: @nastasha-solomon, @approksiu